manubot/cite/tests/test_cite_command.py

```diff
@@ -65,38 +61,75 @@ def test_cite_command_file(tmpdir):
-    if pandoc_version < (2, 0):
-        pytest.skip("Test requires pandoc >= 2.0 to support --lua-filter and --csl=URL")
-    expected = (
-        pathlib.Path(__file__).parent
-        .joinpath('cite-command-rendered', expected)
-        .read_text()
-    )
-    args = [
-        'manubot', 'cite', '--render',
-        '--csl', 'https://github.com/greenelab/manubot-rootstock/raw/e83e51dcd89256403bb787c3d9a46e4ee8d04a9e/build/assets/style.csl',
-        'arxiv:1806.05726v1', 'doi:10.7717/peerj.338', 'pmid:29618526',
-    ] + args
-    process = subprocess.run(
-        args,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.PIPE,
-        universal_newlines=True,
-    )
-    print(shlex_join(process.args))
-    print(process.stdout)
-    print(process.stderr)
-    assert process.stdout == expected
+class Base_cite_command_render_stdout():
+    @classmethod
+    def expected_output(cls, filename):
+        return (
+            pathlib.Path(__file__).parent
+            .joinpath('cite-command-rendered', filename)
+            .read_text()
+        )
+
+    @classmethod
+    def render(self, format_args):
+        args = [
+            'manubot',
+            'cite',
+            '--render',
+            '--csl',
+            'https://github.com/greenelab/manubot-rootstock/raw/e83e51dcd89256403bb787c3d9a46e4ee8d04a9e/build/assets/style.csl',
+            'arxiv:1806.05726v1',
+            'doi:10.7717/peerj.338',
+            'pmid:29618526',
+        ] + format_args
+        process = subprocess.run(
+            args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+        )
+        print(shlex_join(process.args))
+        print(process.stdout)
+        print(process.stderr)
+        return process.stdout
+
+
+@pytest.mark.skipif(
+    pandoc_version < (2, 0),
+    reason="Test requires pandoc >= 2.0 to support --lua-filter and --csl=URL")
+class Test_cite_command_render_stdout_above_pandoc_v2(
+        Base_cite_command_render_stdout):
+    def test_no_arg(self):
+        assert self.render([]) == \
+            self.expected_output('references-plain.txt')
+
+    def test_plain(self):
+        assert self.render(['--format', 'plain']) == \
+            self.expected_output('references-plain.txt')
+
+
+@pytest.mark.skipif(
+    pandoc_version < (2, 5),
+    reason=("Testing markdown, html or jats formats "
+            "assumes pandoc >= 2.5")
+)
+class Test_cite_command_render_stdout_above_pandoc_v2_5(
+        Base_cite_command_render_stdout):
+    def test_markdown(self):
+        assert self.render(['--format', 'markdown']) == \
+            self.expected_output('references-markdown.md')
+
+    def test_html(self):
+        assert self.render(['--format', 'html']) == \
+            self.expected_output('references-html.html')
+
+    def test_jats(self):
+        if pandoc_version >= (2, 7, 3):
+            filename = 'references-jats-2.7.3.xml'
```

Comment:
Would be best to have no hardcoded version paths in this test file.

Assume the following:

1. We will frequently be updating the Pandoc version used for CI.
2. Everytime we update pandoc, all outputs (not just JATS) have a strong probability of breaking.

Therefore, I think it's best if we have the tests detect the pandoc version and look for the appropriate output. But we want to avoid having to change the test everytime we update the pandoc version and create new outputs. Does that make sense?

Reply:
> have no hardcoded version paths in this test file

> we have the tests detect the pandoc version and look for the appropriate output.

Not sure what you mean - what kind of output the tests would need to detect? It must be a straightforward idea, but I'm not catching it.

Reply:
Something that works for all formats / versions and doesn't require updating `test_cite_command.py` with special conditions. Something along the lines of:

```python
# Something like this to generate filename
filename = f"references-{format}-{'.'.join(pandoc_version)}.{extension}"
if not filename.exists():
    assert 'CI' not in os.environ
    # warning like "please create {filename} for test to proceed"
    pytest.skip(...)
# Proceed with filename.read_text and test
```

My experience has been that more often than not, pandoc updates do change the output, so we might as well assume they always do and only evaluate if the version is correct. Does that make sense?

I don't think it's a huge deal if the test is skipped locally... but we could issue a warning so the user is aware.

Reply:
Here is a version of the logic that may be compatible with your suggestion as I understand it: 
1. we are sure about pandoc version in on the CI, because we have installed it there and produced the output files with that version
2. producing new output files should not touch the code of tests and be a change of artefacts, not the code
3. because we are never sure of local pandoc setting we can sacrifice the strengths on local tests

There is a cost:
- with `f"references-{format}-{'.'.join(pandoc_version)}.{extension}"` we accrue the complexity of naming (means we might need special tools/scripts to make those files). 

Also on subjectively on implementation:
- ideally, a test should be ambiguious, mixing warnigns and test results sends a noisy signal
- `assert 'CI' not in os.environ` is tempting to write, but it is a shorthand, because assert may be [disabled by `-o`  flag](https://stackoverflow.com/questions/1273211/disable-assertions-in-python)

Suggestion: if we can to make all output file names contain a version, as you proposed, maybe there can be a rather simple rule for a test:
- find output file that matches installed pandoc version
- if found a match, run test on it
   - if success, move on to next test
   - if test failed - fail
- if no version matched - try other output file versions (there is probably just one)

So, if we keep `reference-jats-2.7.3.txt` and the rest of `reference-{format}-2.7.2.txt` files in the repo, both remote and local versions would work. This would also simplify the test lookup.

Reply:
As for `assert`s I've encountered its use inside code outside tests, that's why I pointed to the usage in tests:  

https://github.com/manubot/manubot/blob/cf9564dd4a31701e48a487ecd716b2cea6aca53a/manubot/cite/csl_item.py#L43

One can think of `assert` vs throughing exceptions as matter of style, but I think exceptions are cleaner, based on [this](https://stackoverflow.com/questions/944592/best-practice-for-python-assert) and [this](https://stackoverflow.com/questions/42837054/are-assert-methods-acceptable-in-production), that is why I called `assert` a shorthand, something you write in place of a longer, but more stable expression. 

In your test example the asset was not related to code under test, so it seemed a shorthand too.